Mathlib/LinearAlgebra/TensorProduct/Tower.lean

```diff
@@ -176,27 +183,210 @@ nonrec def mk : M →ₗ[A] N →ₗ[R] M ⊗[R] N :=
+    add_apply, add_tmul]
+
+theorem map_add_right (f : M →ₗ[A] P) (g₁ g₂ : N →ₗ[R] Q) :
+    map f (g₁ + g₂) = map f g₁ + map f g₂ := by
+  ext
+  simp_rw [curry_apply, TensorProduct.curry_apply, restrictScalars_apply, add_apply, map_tmul,
+    add_apply, tmul_add]
+
+theorem map_smul_right (r : R) (f : M →ₗ[A] P) (g : N →ₗ[R] Q) : map f (r • g) = r • map f g := by
+  ext
+  simp_rw [curry_apply, TensorProduct.curry_apply, restrictScalars_apply, smul_apply, map_tmul,
+    smul_apply, tmul_smul]
+
+theorem map_smul_left (r : B) (f : M →ₗ[A] P) (g : N →ₗ[R] Q) : map (r • f) g = r • map f g := by
+  ext
+  simp_rw [curry_apply, TensorProduct.curry_apply, restrictScalars_apply, smul_apply, map_tmul,
+    smul_apply, smul_tmul']
+
+variable (R A B M N P Q)
+
+/-- Heterobasic version of `TensorProduct.map_bilinear` -/
+def mapBilinear : (M →ₗ[A] P) →ₗ[B] (N →ₗ[R] Q) →ₗ[R] (M ⊗[R] N →ₗ[A] P ⊗[R] Q) :=
+  { toFun := fun f =>
+    { toFun := fun g => map f g
+      map_add' := fun _g₁ _g₂ => map_add_right _ _ _
+      map_smul' := fun _c _g => map_smul_right _ _ _ }
+    map_add' := fun _f₁ _f₂ => LinearMap.ext <| fun _g => map_add_left _ _ _
+    map_smul' := fun _c _f => LinearMap.ext <| fun _g => map_smul_left _ _ _ }
+
+variable {R A B M N P Q}
+
+@[simp]
+theorem mapBilinear_apply (f : M →ₗ[A] P) (g : N →ₗ[R] Q) :
+    mapBilinear R A B M N P Q f g = map f g :=
+  rfl
+
+variable (R A B M N P Q)
+
+/-- Heterobasic version of `TensorProduct.homTensorHomMap` -/
+def homTensorHomMap : ((M →ₗ[A] P) ⊗[R] (N →ₗ[R] Q)) →ₗ[B] (M ⊗[R] N →ₗ[A] P ⊗[R] Q) :=
+  lift <| mapBilinear R A B M N P Q
+
+variable {R A B M N P Q}
+
+@[simp] theorem homTensorHomMap_apply (f : M →ₗ[A] P) (g : N →ₗ[R] Q) :
+    homTensorHomMap R A B M N P Q (f ⊗ₜ g) = map f g :=
+  rfl
+
+/-- Heterobasic version of `TensorProduct.congr` -/
+def congr (f : M ≃ₗ[A] P) (g : N ≃ₗ[R] Q) : (M ⊗[R] N) ≃ₗ[A] (P ⊗[R] Q) :=
+  LinearEquiv.ofLinear (map f g) (map f.symm g.symm)
+    (ext fun m n => by simp)
+    (ext fun m n => by simp)
+
+@[simp] theorem congr_tmul (f : M ≃ₗ[A] P) (g : N ≃ₗ[R] Q) (m : M) (n : N) :
+    congr f g (m ⊗ₜ n) = f m ⊗ₜ g n :=
+  rfl
+
+@[simp] theorem congr_symm_tmul (f : M ≃ₗ[A] P) (g : N ≃ₗ[R] Q) (p : P) (q : Q) :
+    (congr f g).symm (p ⊗ₜ q) = f.symm p ⊗ₜ g.symm q :=
+  rfl
+
+end Semiring
+
+section CommSemiring
+
+variable [CommSemiring R] [CommSemiring A] [Semiring B] [Algebra R A] [Algebra R B]
+
+variable [AddCommMonoid M] [Module R M] [Module A M] [Module B M]
+variable [IsScalarTower R A M] [IsScalarTower R B M] [SMulCommClass A B M]
+
+variable [AddCommMonoid N] [Module R N]
+
+variable [AddCommMonoid P] [Module R P] [Module A P] [Module B P]
+variable [IsScalarTower R A P] [IsScalarTower R B P] [SMulCommClass A B P]
+
+variable [AddCommMonoid Q] [Module R Q]
+
+variable (R A B M N P Q)
+
 attribute [local ext high] TensorProduct.ext
 
+section assoc
+variable [Algebra A B] [IsScalarTower A B M]
```

Comment:
It genuinely does my head in that you don't need `IsScalarTower R A B` here. In the kind of mathematics I'm used to, I'm not sure it would ever happen that A and B are R-algebras and B is an A-algebra and the diagram doesn't commute. However I am not used to non-commutative ring theory and I see that B isn't commutative.